daml-script/dump/src/main/scala/com/daml/script/dump/TreeUtils.scala

```diff
@@ -181,29 +187,47 @@ object TreeUtils {
   sealed trait Command
   final case class CreateCommand(createdEvent: CreatedEvent) extends Command
   final case class ExerciseCommand(exercisedEvent: ExercisedEvent) extends Command
+  final case class ExerciseByKeyCommand(
+      exercisedEvent: ExercisedEvent,
+      templateId: Identifier,
+      contractKey: Value,
+  ) extends Command
   final case class CreateAndExerciseCommand(
       createdEvent: CreatedEvent,
       exercisedEvent: ExercisedEvent,
   ) extends Command
 
   object Command {
     def fromTree(tree: TransactionTree): Seq[Command] = {
+      val contractKeys = mutable.HashMap.empty[ContractId, Value]
       val rootEvents = tree.rootEventIds.map(tree.eventsById(_).kind)
       val commands = ListBuffer.empty[Command]
       rootEvents.foreach {
         case Kind.Empty =>
         case Kind.Created(createdEvent) =>
+          createdEvent.contractKey.foreach { contractKey =>
+            contractKeys += ContractId(createdEvent.contractId) -> contractKey
+          }
           commands += CreateCommand(createdEvent)
         case Kind.Exercised(exercisedEvent) =>
-          commands.lastOption match {
-            case Some(CreateCommand(createdEvent))
+          lazy val optCreateAndExercise = commands.lastOption.flatMap {
```

Comment:
seems like we always force this so not sure it needs to be lazy.

Reply:
True, removed.